app/controllers/shared/api.php

```diff
@@ -19,234 +19,267 @@
-                throw new Exception('Invites authentication is disabled for this project', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+            $abuse = new Abuse($timeLimit);
+
+            if ($timeLimit->limit() && ($timeLimit->remaining() < $closestLimit || is_null($closestLimit))) {
+                $closestLimit = $timeLimit->remaining();
+                $response
+                    ->addHeader('X-RateLimit-Limit', $timeLimit->limit())
+                    ->addHeader('X-RateLimit-Remaining', $timeLimit->remaining())
+                    ->addHeader('X-RateLimit-Reset', $timeLimit->time() + $route->getLabel('abuse-time', 3600))
+                ;
             }
-            break;
 
-        case 'jwt':
-            if (($auths['JWT'] ?? true) === false) {
-                throw new Exception('JWT authentication is disabled for this project', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
+            if (
+                (App::getEnv('_APP_OPTIONS_ABUSE', 'enabled') !== 'disabled' // Route is rate-limited
+                && $abuse->check()) // Abuse is not disabled
+                && (!$isAppUser && !$isPrivilegedUser)
+            ) { // User is not an admin or API key
+                throw new Exception('Too many requests', 429, Exception::GENERAL_RATE_LIMIT_EXCEEDED);
             }
-            break;
+        }
 
-        default:
-            throw new Exception('Unsupported authentication route', 501, Exception::USER_AUTH_METHOD_UNSUPPORTED);
-            break;
-    }
-}, ['utopia', 'request', 'project'], 'auth');
+        /*
+        * Background Jobs
+        */
+        $events
+            ->setEvent($route->getLabel('event', ''))
+            ->setProject($project)
+            ->setUser($user)
+        ;
+
+        $mails
+            ->setProject($project)
+            ->setUser($user)
+        ;
+
+        $audits
+            ->setMode($mode)
+            ->setUserAgent($request->getUserAgent(''))
+            ->setIP($request->getIP())
+            ->setEvent($route->getLabel('event', ''))
+            ->setProject($project)
+            ->setUser($user)
+        ;
 
-App::shutdown(function (App $utopia, Request $request, Response $response, Document $project, Event $events, Audit $audits, Stats $usage, Delete $deletes, EventDatabase $database, string $mode, Database $dbForProject) {
+        $usage
+            ->setParam('projectId', $project->getId())
+            ->setParam('httpRequest', 1)
+            ->setParam('httpUrl', $request->getHostname() . $request->getURI())
+            ->setParam('httpMethod', $request->getMethod())
+            ->setParam('httpPath', $route->getPath())
+            ->setParam('networkRequestSize', 0)
+            ->setParam('networkResponseSize', 0)
+            ->setParam('storage', 0)
+        ;
+
+        $deletes->setProject($project);
+        $database->setProject($project);
+    });
+
+App::init()
+    ->groups(['auth'])
```

Comment:
The rest of PR puts groups under injection. Let's do it here too for consistency.